travelpal/lib/travelpal_web/controllers/weather_controller.ex

```diff
@@ -1,26 +1,90 @@
 defmodule TravelpalWeb.WeatherController do
   use TravelpalWeb, :controller
 
-  # @TODO decide if other functions are needed
-  def weather_url, do: "https://query.yahooapis.com/v1/public/yql"
+  alias Travelpal.ExternalAPI
+  alias Travelpal.ExternalAPI.Weather
 
-  def get_weather_by_city(conn, %{"city" => city}) do
+  action_fallback TravelpalWeb.FallbackController
+
+  def index(conn, _params) do
+    render(conn, "index.json")
+  end
+
+  # Does a search for weather data on the specified city from either the database or an external API request and
+  # returns the results in JSON format.
+  def search(conn, %{"city" => city}) do
+    # @TODO BUG: the city parameter only supports single-word cities, so requested multi-word cities will always
+    # return nil when existing data is searched for in the database. Not a huge deal since it still has the external
+    # request as a fallback.
+
+    city = String.capitalize(city)
+    # weather data for the city in the database; returns a Weather struct with data or nil if none exists
+    existing_data = ExternalAPI.get_weather_by_city(city)
+    # if there exists a record for the city then use the existing data, else make a request for new data
+    weather = if (existing_data != nil), do: existing_data, else: request_weather_by_city(city)
+
+    if (existing_data != nil) do
+      # updates the weather in the database with new external data if it's outdated
+      if (Date.compare(weather.date, Date.utc_today()) != :eq) do
+        weather = request_weather_by_city(city)
+        ExternalAPI.update_weather(existing_data, weather)
+      end
+    else
+      # adds a weather record for the city in the database
+      ExternalAPI.create_weather(weather)
+    end
+
+    render(conn, "show.json", weather: weather)
+  end
+
+  # Makes an external API call to request Yahoo Weather data for a specified city.
+  defp request_weather_by_city(city) do
+    weather_url = "https://query.yahooapis.com/v1/public/yql"
     # relevant columns from the weather table
-    columns = ["units", "location", "item"]
+    columns = ["units", "location", "item",]
     |> Enum.join(", ")
     # Yahoo Weather API uses YSQL to specify data
     ysql_query = "SELECT #{columns} FROM weather.forecast WHERE woeid in (SELECT woeid FROM geo.places(1) WHERE text=\"#{city}\")"
-    uri = URI.encode(weather_url() <> "?q=#{ysql_query}&format=json")
+    uri = URI.encode(weather_url <> "?q=#{ysql_query}&format=json")
     # comment out HTTP request for dev purposes and use dummy data instead
     #res = HTTPoison.get!(uri)
-    res = dummy_data()
-    data = Poison.decode!(res.body)
+    res = dummy_data(city)
```

Comment:
I'm getting an `undefined function dummy_data/1` error because you're trying to pass city as an argument, but `dummy_data()` below takes no arguments.